### ISSUE TITLE
failed uploads are not tried again next time

### ISSUE THREAD

**Issue by PontiacCZ** [mendhak/gpslogger#622]
I have set up the GPS Logger to automatically upload the GPX when I finish logging. A few day ago I forgot to turn on cellular data so the upload failed. The GPS logger did not let me know but I understand that since it probably uploads the data in the background after I  exit the application.

BUT I would expect it to try the failed upload next time I launch the application or at least give me a clue that I need to do it manually. Neither of those actions occurred. I happened to find out about the missing tracklog when I opened the OSM page with my traces. I then uploaded the GPX manually.

So is it possible to update this great app so that it attempts to upload the GPX next time when it fails?

**Comment by mendhak** [comment]
I've just tried this a few times - logging without connectivity, exiting the app.  Then restore connectivity and open the app.  At that point the pending uploads happened.  The behavior that you're expecting is what should be happening.  

So can this be an isolated error - do you see it fail all the time?  If the failure is consistent - have a look at the Log View and sometimes you may see the corresponding error there.  OR, try enabling debug logging and that writes out all the details and when the next upload fails the error will be in there.  However the debug log creates a large file, which means you shouldn't keep debug logging always enabled.

**Comment by PontiacCZ** [comment]
Try to be on wi-fi without network connectivity and disable cell data as well. 

Why I am saying this: there's one thing new in my car - an on-board camera using wi-fi (DDPai M4). The camera application unfortunately has sort of a routing bug  - when the LTE cellular data are enabled, it won't connect to the camera. That's why I sometimes disable cellular data to be able to use the app and both cases of failed uploads that I know about so far happened in this scenario - the phone was connected via wi-fi to the camera and LTE data were off. 

Will turn on the debug logging and do another attempts to reproduce the issue.

I have Nexus 6 with Android 7.1.1.

**Comment by mendhak** [comment]
I've not been able to recreate a wifi network without network connectivity.  I did attempt to unplug my router ethernet, but the current algorithm still worked just fine - it waited until network connectivity was restored before proceeding to make the HTTP requests. 

So I'm having to make a guess here for your case.  I've now set [this line to isConnected](https://github.com/mendhak/gpslogger/commit/28b56ee94a7a6d8d1638a2f113238b97e757a7f9#diff-3373980733513c4410a4ab5e8d7f9a7dL71).  That should wait for a more established connection before allowing HTTP jobs to continue.

You can try to sideload these APKs to see if it makes a difference in that setup

[APK](https://drive.google.com/file/d/1Gog9abQXnpgfk26lIvnv5xkzQNMXKQkq/view?usp=sharing)
[signature](https://drive.google.com/file/d/1Oi46G_KN2AmSMEfZ8huk4ivfY3qEUDnW/view?usp=sharing)
[checksum](https://drive.google.com/file/d/1vW_l3A9xekfkPu6wF668Unu6mnZpfnGT/view?usp=sharing)

Failing that your debug log would be more useful to figure out what's wrong.

**Comment by PontiacCZ** [comment]
First try to analyze the debug log I have just sent you.

Two files failed to upload today:
20180228161054.gpx - I intentionally turned off the data service so it could not upload
20180228161204.gpx - data was on again but I ended my ride in a deep valey with no cellular signal at all

Today's last GPX, 20180228175328.gpx, got uploaded well as I was again in civilized world with LTE coverage. I expectedt GPSlogger to use this opportunity to retry uploading the previous two files which did not happen.

The question is why...

**Comment by mendhak** [comment]
The error log was useful - so what I'm seeing is, the jobs are indeed queued.  When a network connection becomes available, the job is tried.  However the DNS is not resolving.  So it tries to find `www.openstreetmap.org` but the network cannot provide anything.  At that point it tries a few times and fails the job. 

In the APK above, I've used a [slightly different method to test connectivity](https://developer.android.com/reference/android/net/NetworkInfo.html#isConnected\(\)).  You can try that one.  Ultimately it's coming down to - the OS is reporting that a network is available and ready to use, if this `isConnected` provides wrong info then I don't believe there is a solution for this unfortunately. Previously I was using `isConnectedOrConnecting` and I'm hoping the difference is whether DNS resolution is possible or not, as in, there's a tested path.

	16:11:20 ERROR AppSettings$1.e:96 - error while executing job com.mendhak.gpslogger.senders.osm.OSMJob@7860a66
	java.net.UnknownHostException: Unable to resolve host "www.openstreetmap.org": No address associated with hostname
		at java.net.Inet6AddressImpl.lookupHostByName(Inet6AddressImpl.java:125) ~[na:0.0]
		at java.net.Inet6AddressImpl.lookupAllHostAddr(Inet6AddressImpl.java:74) ~[na:0.0]
		at java.net.InetAddress.getAllByName(InetAddress.java:752) ~[na:0.0]
		at okhttp3.Dns$1.lookup(Dns.java:39) ~[na:0.0]
		at okhttp3.internal.http.RouteSelector.resetNextInetSocketAddress(RouteSelector.java:173) ~[na:0.0]
		at okhttp3.internal.http.RouteSelector.nextProxy(RouteSelector.java:139) ~[na:0.0]
		at okhttp3.internal.http.RouteSelector.next(RouteSelector.java:81) ~[na:0.0]
		at okhttp3.internal.http.StreamAllocation.findConnection(StreamAllocation.java:174) ~[na:0.0]
		at okhttp3.internal.http.StreamAllocation.findHealthyConnection(StreamAllocation.java:127) ~[na:0.0]
		at okhttp3.internal.http.StreamAllocation.newStream(StreamAllocation.java:97) ~[na:0.0]
		at okhttp3.internal.http.HttpEngine.connect(HttpEngine.java:289) ~[na:0.0]
		at okhttp3.internal.http.HttpEngine.sendRequest(HttpEngine.java:241) ~[na:0.0]
		at okhttp3.RealCall.getResponse(RealCall.java:240) ~[na:0.0]
		at okhttp3.RealCall$ApplicationInterceptorChain.proceed(RealCall.java:198) ~[na:0.0]
		at se.akerfeldt.okhttp.signpost.SigningInterceptor.intercept(SigningInterceptor.java:46) ~[na:0.0]
		at okhttp3.RealCall$ApplicationInterceptorChain.proceed(RealCall.java:187) ~[na:0.0]
		at okhttp3.RealCall.getResponseWithInterceptorChain(RealCall.java:160) ~[na:0.0]
		at okhttp3.RealCall.execute(RealCall.java:57) ~[na:0.0]
		at com.mendhak.gpslogger.senders.osm.OSMJob.onRun(OSMJob.java:95) ~[na:0.0]
		at com.path.android.jobqueue.Job.safeRun(Job.java:145) ~[na:0.0]
		at com.path.android.jobqueue.JobHolder.safeRun(JobHolder.java:90) ~[na:0.0]
		at com.path.android.jobqueue.executor.JobConsumerExecutor$JobConsumer.run(JobConsumerExecutor.java:306) ~[na:0.0]
		at java.lang.Thread.run(Thread.java:761) ~[na:0.0]
	Caused by: android.system.GaiException: android_getaddrinfo failed: EAI_NODATA (No address associated with hostname)
		at libcore.io.Posix.android_getaddrinfo(Native Method) ~[na:0.0]
		at libcore.io.ForwardingOs.android_getaddrinfo(ForwardingOs.java:55) ~[na:0.0]
		at java.net.Inet6AddressImpl.lookupHostByName(Inet6AddressImpl.java:106) ~[na:0.0]
		... 22 common frames omitted
	16:11:20 ERROR OSMJob.shouldReRunOnThrowable:116 - Could not send to OpenStreetMap
	java.net.UnknownHostException: Unable to resolve host "www.openstreetmap.org": No address associated with hostname
		at java.net.Inet6AddressImpl.lookupHostByName(Inet6AddressImpl.java:125) ~[na:0.0]
		at java.net.Inet6AddressImpl.lookupAllHostAddr(Inet6AddressImpl.java:74) ~[na:0.0]
		at java.net.InetAddress.getAllByName(InetAddress.java:752) ~[na:0.0]
		at okhttp3.Dns$1.lookup(Dns.java:39) ~[na:0.0]
		at okhttp3.internal.http.RouteSelector.resetNextInetSocketAddress(RouteSelector.java:173) ~[na:0.0]
		at okhttp3.internal.http.RouteSelector.nextProxy(RouteSelector.java:139) ~[na:0.0]
		at okhttp3.internal.http.RouteSelector.next(RouteSelector.java:81) ~[na:0.0]
		at okhttp3.internal.http.StreamAllocation.findConnection(StreamAllocation.java:174) ~[na:0.0]
		at okhttp3.internal.http.StreamAllocation.findHealthyConnection(StreamAllocation.java:127) ~[na:0.0]
		at okhttp3.internal.http.StreamAllocation.newStream(StreamAllocation.java:97) ~[na:0.0]
		at okhttp3.internal.http.HttpEngine.connect(HttpEngine.java:289) ~[na:0.0]
		at okhttp3.internal.http.HttpEngine.sendRequest(HttpEngine.java:241) ~[na:0.0]
		at okhttp3.RealCall.getResponse(RealCall.java:240) ~[na:0.0]
		at okhttp3.RealCall$ApplicationInterceptorChain.proceed(RealCall.java:198) ~[na:0.0]
		at se.akerfeldt.okhttp.signpost.SigningInterceptor.intercept(SigningInterceptor.java:46) ~[na:0.0]
		at okhttp3.RealCall$ApplicationInterceptorChain.proceed(RealCall.java:187) ~[na:0.0]
		at okhttp3.RealCall.getResponseWithInterceptorChain(RealCall.java:160) ~[na:0.0]
		at okhttp3.RealCall.execute(RealCall.java:57) ~[na:0.0]
		at com.mendhak.gpslogger.senders.osm.OSMJob.onRun(OSMJob.java:95) ~[na:0.0]
		at com.path.android.jobqueue.Job.safeRun(Job.java:145) ~[na:0.0]
		at com.path.android.jobqueue.JobHolder.safeRun(JobHolder.java:90) ~[na:0.0]
		at com.path.android.jobqueue.executor.JobConsumerExecutor$JobConsumer.run(JobConsumerExecutor.java:306) ~[na:0.0]
		at java.lang.Thread.run(Thread.java:761) ~[na:0.0]
	Caused by: android.system.GaiException: android_getaddrinfo failed: EAI_NODATA (No address associated with hostname)
		at libcore.io.Posix.android_getaddrinfo(Native Method) ~[na:0.0]
		at libcore.io.ForwardingOs.android_getaddrinfo(ForwardingOs.java:55) ~[na:0.0]
		at java.net.Inet6AddressImpl.lookupHostByName(Inet6AddressImpl.java:106) ~[na:0.0]
		... 22 common frames omitted
	16:11:20 DEBUG AppSettings$1.d:91 - safeRunResult for com.mendhak.gpslogger.senders.osm.OSMJob@7860a66 : false. re run:false. cancelled: false
	16:11:20 DEBUG OSMJob.onCancel:111 - OSM Job cancelled 
	16:11:20 DEBUG GpsMainActivity.onEventMainThread:1253 - OSM Event completed, success: false
	16:11:20 ERROR GpsMainActivity.onEventMainThread:1257 - OpenStreetMap-Could not upload the file

**Comment by mendhak** [comment]
I've included this in the v93 release anyway, closing this now but reopen it if needed.

**Comment by PontiacCZ** [comment]
So to inform you: after I updated to v93 today I went out geocaching. I did two very short tracklogs first with cellular data off. I am now looking at my traces on openstreetmap and those two are not there (subsequent three are uploaded). :-(

Failed GPX's:
20180310160627.gpx
20180310160757.gpx

Sending you the current debuglog now. I understand you won't be able to do anything more about it but anyway. 

I'll need to make sure in the future that I am connected to the internet or if not, then upload the files later manually. :-) Thank you very much for your effort, I will continue to use GPSlogger as it is very reliable and configurable application.
